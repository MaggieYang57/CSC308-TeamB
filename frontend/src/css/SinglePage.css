.header{
    display: flex;
    justify-content: left;
    margin-top: 100px;
}
.header h1{
    margin-left: 200px;
    font-family: 'Mulisih', sans-serif;
    font-size: 30px;
    font-weight: 800;
    color:#265b5f;
}
.header h2{
    margin-left: 50px;
    font-size: 25px;
    color:#265b5f;
}

h3{
    justify-content: center;
    text-align: center;
    font-family: 'Mulish', sans-serif;
    font-size: 25px;
    color:#265b5f;
}

.boxed {
    box-sizing: content-box;
    width: 100%;
    border: solid black 1px;
    padding: 15px;
    border-radius: 25px;
}

img{
    display: flex;
    justify-content: left;
    margin-left: 200px;
    border-radius: 50px;
}

/****************side information*******************/
.single-info{
    text-align: left;
    justify-content: left;
    margin-top: -300px;
    margin-left: 670px;
    margin-right: 200px;
    font-family: 'Quicksand', sans-serif;
}

#desc
{
    margin: 0;
    font-weight:500;
    font-family: 'Quicksand', sans-serif;
    font-size: 15px;
}

.stats{
    display: flex;
    align-items: center;
    justify-content: space-evenly;
}
.difficulty{
    font-size: 25px;
    box-sizing: border-box;
    padding: 2px 10px;
    border: solid;
    border-width: 3px;
    border-radius: 10px;
    border-color: #10A5F5;
    color: #10A5F5;
}
#rating{
    display: flex;
    color: #ff8b3d;
}
#rating-num{
    display: flex;
    color: #ff8b3d;
}

.rate-me h3
{
    display: flex;;
    justify-content: space-between;
    margin-right: 1vw;
    font-size: 30px;
    color:#265b5f;
}

/************************Rating**************************/

#post-rating
{
    -webkit-appearance: none;
    display: inline-block;
    padding: 0.75rem 1.25rem;
    margin: 0 0 0.5rem 1vw;
    vertical-align: middle;

    font-weight: 600;
    font-family: Helvetica Neue, Helvetica, Arial, sans-serif;
    font-size: 1rem;
    text-transform: none;
    text-align: center;
    text-decoration: none;
    border: 1px solid #0366ee;
    border-radius: 10px;
    background: #0366ee;
    color: #ffffff;
    
    cursor: pointer;
    line-height: 1;
}

#post-rating:hover{
    border: 1px solid #0250bc;
    background: #0250bc;
    color: #ffffff;
    text-decoration: none;
  }
  
  #post-rating:focus{
    border: 1px solid #0250bc;
    background: #0250bc;
    color: #ffffff;
    text-decoration: none;
  }

/**********************Reviews********************/

.reviews{
    display: flex;
    justify-content: left;
    margin-top: 1vw;
}
.reviews h2{
    display: flex;
    justify-content: left;
    margin-left: 200px;
    font-size: 25px;
    font-weight: 800;
    color:#265b5f;
}
