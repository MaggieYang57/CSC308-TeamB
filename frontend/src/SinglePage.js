import React from "react";
import "./css/SinglePage.css";
import { withRouter } from "react-router-dom";
import { PropTypes } from 'prop-types';

import WeatherWidget from "./components/singlepage/WeatherWidget";
import ReviewTable from "./components/ReviewTable";
import SaveButton from "./components/singlepage/SaveButton"

const averageRatings = (ratings) => {
  let sum = 0;
  for (const i in ratings) sum += +ratings[i];
  return (sum / ratings.length).toFixed(1);
};

class SinglePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount() {
    fetch("http://localhost:3001/hike/" + this.props.match.params.id)
      .then((res) => res.json())
      .then((data) => {
        console.log("data", data);
        this.setState({ ...data[0] });
        document.getElementById("rating-num").innerText = averageRatings(
          this.state.rating
        );
      });
  }

  postRating = () => {
    const ratingSelect = document.getElementById("select-rating");
    const rate = +ratingSelect.options[ratingSelect.selectedIndex].value;
    const data = {
      rating: rate,
    };

    fetch("http://localhost:3001/hike/" + this.state._id + "/rating", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    }).then(() => {
      this.state.rating.push(rate);
      document.getElementById("rating-num").innerText = averageRatings(
        this.state.rating
      );
    });
  };

  /* checkSaved = () => {
    if (localStorage.getItem("isLoggedIn") === "true")
    {
        const hikeID = this.state._id
        const hike = hikeID
        const user = localStorage.getItem("_id")
        const data = {
            id: hike,
            user: user
        };
        fetch("http://localhost:3001/save/" + this.state._id + "/check", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
        })
        .then((res) => {
            if (res.status === 404)
                return false;
            else if (res.status === 200)
                return true;
        })
    }
    else 
      return false;
  }
 */

  render() {
    // const save = this.checkSaved()
    return (
      <div className="hike">
        <div className="header">
          <h1>{this.state.title}</h1>
          <h2>- {this.state.location}</h2>
          <a href={"/review/" + this.state._id}>
            <button id="review-button">Write a Review</button>
          </a>
          <SaveButton hike={this.state.title}/>
        </div>

        <img src={this.state.imagesrc} height="300" width="400" />
        <div className="single-info">
          <div className="stats">
            <h2 className="difficulty">Difficulty: {this.state.difficulty}</h2>
            <h2 id="rating">
              ★<h2 id="rating-num" />
            </h2>
          </div>
          <p id="desc">{this.state.description}</p>
          <br></br>
          <WeatherWidget city={this.state.location} />
          <br></br>
          <p>{this.state.tags}</p>
        </div>
        <div className="rate-me">
          <label id="rating-label" htmlFor="select-rating">
            <h3>Rate Me!</h3>
          </label>
          <select id="select-rating" length="20">
            <option value="5">5 ★</option>
            <option value="4">4 ★</option>
            <option value="3">3 ★</option>
            <option value="2">2 ★</option>
            <option value="1">1 ★</option>
          </select>
          <button id="post-rating" onClick={this.postRating}>
            Post Rating
          </button>
        </div>
        <div className="reviews">
          <h2 className="reviews-title">
            <hr />
            Reviews
            <hr />
          </h2>
          <ReviewTable reviewList={this.state.reviews} />
        </div>
      </div>
    );
  }
}

SinglePage.propTypes = {
  match: PropTypes.object,
};
export default withRouter(SinglePage);
