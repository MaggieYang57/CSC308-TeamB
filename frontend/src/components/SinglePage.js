import React from 'react';
import '../css/SinglePage.css';
import moment from 'moment';

class SinglePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  getWeather() {
    fetch('http://api.openweathermap.org/data/2.5/weather?q='+this.state.location+'&units=imperial&appid=9bb2aad28619e8c45992b312cb325c36')
  .then(res => res.json())
  .then(data => {
    if(data)
    {
     var icon = JSON.stringify(data.weather[0].icon);
     icon = icon.replace("\"","");
     icon = icon.replace("\"","");
    
     document.getElementById("weatherTempLabel").innerHTML = 'Current Temperature: ' + data.main.temp + '&#176; F <br> Feels like: ' + data.main.feels_like + '&#176; F' + '<br> Day Low: ' + data.main.temp_min + '&#176; F <br> Day High: ' + data.main.temp_max + '&#176; F' + '<br>Pressure: ' + data.main.pressure + ' hPa <br> Humidity: ' + data.main.humidity + '%';
     document.getElementById("weatherDescription").innerHTML = data.weather[0].main + ': ' + data.weather[0].description;
     document.getElementById("windDescription").innerHTML = 'Wind: Speed: ' + data.wind.speed + ' mph Direction: ' + data.wind.deg + '&#176;';
     document.getElementById("sunriseSunsetDescription").innerHTML = 'Sunrise: ' + moment.unix(data.sys.sunrise).format('LT') + ' Sunset: ' + moment.unix(data.sys.sunset).format('LT');
     document.getElementById("weatherIcon").src = 'http://openweathermap.org/img/wn/' + icon + '@2x.png'
     return data;
     }
     });
  }


  componentDidMount() {
    fetch('http://localhost:3001/hike/')
      .then(res => res.json())
      .then(data => {
        this.setState({ ...data[0] });
        this.getWeather();
      });
  }

  render() {
    return (
        <div className="hike">
            <div className= "header">
                <h1>{this.state.title}</h1>
                <h2>- {this.state.location}</h2>
            </div>
            <img src = "https://www.margarita-adventures.com/wp-content/uploads/2017/02/Cerro_San_Luis.jpg" height="300" />
            <div className = "info">
                <div className = "stats">
                    <h2 className = "difficulty">{this.state.difficulty}</h2>
                    <h2 className = "rating">â˜…{this.state.rating}</h2>
                </div>
                <p className="desc">{this.state.description}</p>
                <br></br>
                <div className = "boxed">
                <h3>Current Weather Conditions</h3>
                <p id="weatherTempLabel"></p>
                <p id="weatherDescription"></p>
                <p id="windDescription"></p>
                <p id="sunriseSunsetDescription"></p>
                <img id="weatherIcon" src = "" height="100" />
                </div>
                <br></br>
                <p>{this.state.tags}</p>
            </div>
            <div className = "reviews">
                <h2>Reviews</h2>
                <p>{this.state.reviews}</p>
            </div>
        </div>
    );
  }
}
export default SinglePage;
