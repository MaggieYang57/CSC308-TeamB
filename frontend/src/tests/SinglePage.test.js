import React from 'react';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import SinglePage from '../SinglePage';
import createFetchSpy from './CreateFetchSpy';

describe('Single Page', () => {

  beforeEach(() => {
    Object.defineProperty(window, "localStorage", {
      value: {
        removeItem: jest.fn(() => null),
        setItem: jest.fn(() => null),
        getItem: (s) => {
          if (s === "isLoggedIn") {
            return "true";
          }
          return null;
        },
      },
      writable: true
    });

    createFetchSpy({}, [
      {
        saved_trails: ["nope"],
        rating: [2.0]
      }
    ]);

    // eslint-disable-next-line react/display-name
    jest.doMock('../components/singlepage/WeatherWidget', () => 'WeatherWidget');
    // eslint-disable-next-line react/display-name
    jest.doMock('../components/ReviewTable', () => 'ReviewTable');

    const p = document.createElement('p');
    p.id = 'rating-num';
    document.body.appendChild(p);
  })

  it('renders correctly', () => {
    const tree = renderer
      .create(<MemoryRouter> <SinglePage /> </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
