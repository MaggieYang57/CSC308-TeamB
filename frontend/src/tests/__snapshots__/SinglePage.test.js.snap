// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Page renders correctly 1`] = `
Array [
  " ",
  <div
    className="hike"
  >
    <div
      className="header"
    >
      <h1 />
      <h2>
        - 
      </h2>
      <a
        href="/login"
      >
        <button
          id="review-button"
        >
          Write a Review
        </button>
      </a>
      <h3>
        Loading
      </h3>
    </div>
    <img
      height="300"
      width="400"
    />
    <div
      className="single-info"
    >
      <div
        className="stats"
      >
        <h2
          className="difficulty"
        >
          Difficulty: 
        </h2>
        <h2
          id="rating"
        >
          ★
          <h2
            id="rating-num"
          />
        </h2>
      </div>
      <p
        id="desc"
      />
      <br />
      <div
        className="weatherwidget"
      >
        <div
          className="boxed"
        >
          <h3>
            Current Weather Conditions
          </h3>
          <img
            height="100"
            id="weatherIcon"
          />
          <p
            id="weatherTempLabel"
          />
          <p
            id="weatherDescription"
          />
          <p
            id="windDescription"
          />
          <p
            id="sunriseSunsetDescription"
          />
        </div>
      </div>
      <br />
      <p />
    </div>
    <div
      className="rate-me"
    >
      <label
        htmlFor="select-rating"
        id="rating-label"
      >
        <h3>
          Rate Me!
        </h3>
      </label>
      <select
        id="select-rating"
        length="20"
      >
        <option
          value="5"
        >
          5 ★
        </option>
        <option
          value="4"
        >
          4 ★
        </option>
        <option
          value="3"
        >
          3 ★
        </option>
        <option
          value="2"
        >
          2 ★
        </option>
        <option
          value="1"
        >
          1 ★
        </option>
      </select>
      <button
        id="post-rating"
        onClick={[Function]}
      >
        Post Rating
      </button>
    </div>
    <div
      className="reviews"
    >
      <h2
        className="reviews-title"
      >
        <hr />
        Reviews
        <hr />
      </h2>
    </div>
  </div>,
  " ",
]
`;
